### DISCUSSION TITLE
Multiple presets overwriting each other

### DISCUSSION THREAD

**Discussion by JorgenKrieger** [tailwindlabs/tailwindcss#9907] [Help] [answered]
I'm trying to split a config into multiple presets as mentioned in the [Tailwind docs](https://tailwindcss.com/docs/presets#extending-multiple-presets), but as the title of this post indicates, they are overwriting each other. 

The documentation mentions if both configurations contain a color palette (without using `extend`) the latest loaded preset would be used. Understandable, but in my case it's overwriting everything set in the theme object of previous presets. Is this intended behavior or a bug?

<details>
<summary><h3>Example</h3></summary>
I made a small test with the following setup. In this scenario, only the colors would be loaded and the font size would be ignored. If I switch the 2 files around in <code>tailwind.config.js</code> the font size would load but the color wouldn't be usable.

**font.js**
```js
module.exports = {
  theme: {
    fontSize: {
      huge: '10px'
    }
  }
}
```

**colors.js**
```js
module.exports = {
  theme: {
    colors: {
      red: "#F00"
    }
  }
}
```

**tailwind.config.js**
```js
...
  presets: [
    require('./font.js'),
    require('./colors.js'),
  ],
...
```
</details>

**Comment by wongjn** [comment, marked as answer]
I believe this is intended behavior. See my explanation below:

## `theme.<key>` vs `theme.extend.<key>`

Firstly, when you define `theme.<key>`, it completely overrides anything "before" it, whereas defining `theme.extend.<key>` attempts to merge the values to any existing values of `theme.<key>` before it[^1]:
```js
// If we merge the following:
const a = {
  theme: {
    foo: {
      1: 1,
      2: 2,
      3: 3,
    },
  },
};

const b = {
  theme: {
    foo: {
      bar: 'bar',
    },
  },
};

// We get:
const mergedAB = {
  theme: {
    foo: {
      // `foo` values from a are completely overriden.
      bar: 'bar',
    },
  },
};

// Compare if we merge a with this:
const c = {
  theme: {
    extend: {
      foo: {
        bar: 'bar',
      },
    },
  },
};

// We get:
const mergedAC = {
  theme: {
    foo: {
      // `foo` values from a still exist.
      1: 1,
      2: 2,
      3: 3,
      // `foo` values from c are merged in.
      bar: 'bar',
    },
  },
};
```

## Default config in `presets`

> By default, presets themselves extend Tailwind’s [default configuration](https://unpkg.com/browse/tailwindcss@%5E3/stubs/defaultConfig.stub.js) just like your own configuration would.[^2]

This means that for presets in your example, they end up looking something like:

### font.js
```js
module.exports = {
  theme: {
    // …
    // … All Tailwind config defaults here… 
    // …
    colors: {
      inherit: 'inherit',
      current: 'currentColor',
      transparent: 'transparent',
      // …
    },
    fontSize: {
      huge: '10px'
    },
    // …
    // … All Tailwind config defaults here… 
    // …
  },
};
```

### colors.js
```js
module.exports = {
  theme: {
    // …
    // … All Tailwind config defaults here… 
    // …
    colors: {
      red: "#F00"
    },
    fontSize: {
      xs: ['0.75rem', { lineHeight: '1rem' }],
      sm: ['0.875rem', { lineHeight: '1.25rem' }],
      base: ['1rem', { lineHeight: '1.5rem' }],
      // …
    },
    // …
    // … All Tailwind config defaults here… 
    // …
  },
};
```

## The final merge

So when we define these two as presets in our own configuration:
```js
module.exports = {
  presets: [
    require('./font.js'),
    require('./colors.js'),
  ],
};
```

We are actually merging objects like this:

```js
const fonts = {
  theme: {
    // …
    colors: {
      inherit: 'inherit',
      current: 'currentColor',
      transparent: 'transparent',
      // …
    },
    fontSize: {
      huge: '10px'
    },
    // …
  },
};

const colors = {
  theme: {
    // …
    colors: {
      red: "#F00"
    },
    fontSize: {
      xs: ['0.75rem', { lineHeight: '1rem' }],
      sm: ['0.875rem', { lineHeight: '1.25rem' }],
      base: ['1rem', { lineHeight: '1.5rem' }],
      // …
    },
    // …
  },
};
```
And as per the initial explanation on `theme.<key>` vs `theme.extend.<key>`, when we merge `fonts` and `colors`, the `theme.fontSize` in `colors` will completely override the `theme.fontSize` defined in `fonts`.

## Solutions

### Override only defined `theme` keys

If your intention of these presets is to override the default configuration but merge the `theme` values these presets set, you could [set the `presets` key to an empty array to disable the default configuration that these presets would extend from](https://tailwindcss.com/docs/presets#disabling-the-default-configuration):

#### font.js
```diff
 module.exports = {
+  presets: [],
   theme: {
     fontSize: {
       huge: '10px'
     }
   }
 }
```

#### colors.js
```diff
 module.exports = {
+  presets: [],
   theme: {
     colors: {
       red: "#F00"
     }
   }
 }
```
But this will mean no default configuration will be extended at all, so no other utilities will have values. Thus we would need to manually re-include the default configuration manually in our configuration:

```diff
 module.exports = {
   presets: [
+    require('tailwindcss/defaultconfig'),
     require('./font.js'),
     require('./colors.js'),
   ],
 };
```

### Extend

If you don't mind the default values being there for extra values you define in your presets, you could define your values in `theme.extend` instead of `theme`:

#### font.js
```diff
 module.exports = {
   theme: {
-    fontSize: {
-      huge: '10px'
+    extend: {
+      fontSize: {
+        huge: '10px',
+      }
     }
   }
 }
```

#### colors.js
```diff
 module.exports = {
   theme: {
-    colors: {
-      red: "#F00"
+    extend: {
+      colors: {
+        red: "#F00"
+      }
     }
   }
 }
```

[^1]: https://tailwindcss.com/docs/theme#customizing-the-default-theme
[^2]: https://tailwindcss.com/docs/presets#creating-a-preset

**Reply by JorgenKrieger** [reply]
Awesome explanation! That really helped out. Thanks 😄
